dask_expr/_shuffle.py

```diff
@@ -619,6 +626,18 @@ def operation(df, index, name: str, npartitions: int):
             index = partitioning_index(index, npartitions)
         return df.assign(**{name: index})
 
+    def _combine_similar(self, root: Expr):
+        return self._combine_similar_branches(root, (Filter, Projection))
+
+    def _remove_operations(self, frame, remove_ops, skip_ops=None):
+        expr, ops = super()._remove_operations(frame, remove_ops, skip_ops)
+        if len(ops) > 0 and isinstance(ops[0], list):
+            if sorted(ops[0]) == sorted(self.frame.columns):
```

Comment:
I don't like that part very much, but will revisit after we re-ordered the optimisation logic, this should make this a lot easier